include/st_string.h

```diff
@@ -35,6 +35,23 @@
 #   include <filesystem>
 #endif
 
+#ifdef ST_ENABLE_COREFOUNDATION
+#   include <CoreFoundation/CoreFoundation.h>
+#   ifdef __llvm__
+        asm(".linker_option \"-framework\", \"CoreFoundation\"");
+#   endif
```

Comment:
This is a header-only library, so I'm not sure what that would actually do... if it works, then yeah, it's _way_ better than this hackery

Reply:
The CMakeLists.txt installs a CMake target definition for `string_theory` into `${PREFIX}/lib/cmake/string_theory`. I think this gets picked up when you do `find_package(string_theory)` in another project. Currently that target sets up string_theory's include directory - I assume any `target_link_libraries` would also be represented there.

That said, perhaps string_theory shouldn't be responsible for linking in Core Foundation/Foundation. Practically speaking: the Mac Plasma client needs to convert between `ST::string` and `NSString`, so Plasma on Apple platforms will require string_theory with the Foundation integration enabled. But that doesn't mean that other parts of Plasma should be force-linked against Foundation just because they use string_theory (but don't actually depend on the Foundation integration).

Hmm, actually the same thing also applies at the header level... With the Core Foundation integration enabled, anything that includes `<string_theory/string>` will implicitly include all of Core Foundation, even where the integration isn't used. I think this is unavoidable though - because the integration adds new methods to `ST::string`, it must be enabled consistently across all files to avoid violating the one definition rule. The header pollution still isn't nice though :/

Reply:
Hypothetically, I could typedef `CFStringRef` and `NSString` in the main header and declare the functions, but implement them (and include CoreFoundation) in a different optional header file (with the linker options there).

In the cast of Plasma, the CoreLib library already links CoreFoundation so we're not gaining/losing anything there, but string_theory is used in several projects beyond Plasma and I'd like to be as unobtrusive as possible.

The other concern with `target_link_libraries` is that it requires using CMake to have that linking happen, rather than just finding/using the header files as installed to `/usr/local/include` or something. On the other hand, documenting "Hey, you need to link against CoreFoundation (or Foundation, in the case of NSString) if you're using this specific functionality" doesn't seem like the end of the world. If you're in a project where you have CFStringRefs that you want to convert to ST::strings, you probably already need CoreFoundation linked.

Reply:
I am opposed to bad UX such that you have to include certain special things to not blow up your build. It is proper modern CMake to link against all libraries required to consume your project. Therefore, it is imperative that when the Apple string feature is enabled that all required libraries (so called Foundation, from what I can tell) be linked. string_theory is already established as a heavy hitter by its usage of `type_traits`, so you should know what you're getting into by using this library. If the complexity of Foundation is not needed or desired, then the feature should simply be turned off.

Reply:
Fair enough. That would make this integration only really useful for pure Objective-C(++) codebases though, i. e. not Plasma. Objective-C code links against Foundation no matter what (because it's the language's standard library). But for plain C/C++ code, Foundation is a useless dependency, because the API is only designed for Objective-C. Core Foundation *is* useful to C code, but will only be used in Apple-specific code, so it will also be a pointless dependency in e. g. platform-neutral command line tools.

You're right that the proper solution would be to turn this feature off where not needed, but with it being a build-time option on string_theory, it's impossible to do that selectively for e. g. only some Plasma subprojects (let alone having those subprojects interact safely, because ODR).